File: apple/packages/EnteComponents/Sources/EnteComponents/ActionButton.swift
import SwiftUI

public enum ActionVariant: Sendable {
    case primary
    case secondary
    case neutral
    case critical
    case criticalText
    case link
}

public enum ActionDensity: Sendable {
    case regular
    case compact
}

public enum ActionSize: Sendable {
    case small
    case large
}

public struct ActionButton: View {
    @Environment(\.isEnabled) private var enabled
    private let title: String
    private let variant: ActionVariant
    private let size: ActionSize
    private let density: ActionDensity
    private let loading: Bool
    private let action: () -> Void

    public init(
        _ title: String,
        variant: ActionVariant = .primary,
        size: ActionSize = .large,
        density: ActionDensity = .regular,
        loading: Bool = false,
        action: @escaping () -> Void
    ) {
        self.title = title
        self.variant = variant
        self.size = size
        self.density = density
        self.loading = loading
        self.action = action
    }

    public var body: some View {
        let inlineLink = variant == .link && size == .small
        Button(action: action) {
            Text(title)
                .underline(variant == .criticalText || variant == .link)
                .lineLimit(2)
                .font(density == .regular ? EnteTypography.bodyBold : EnteTypography.body)
                .opacity(loading ? 0 : 1)
                .overlay {
                    if loading {
                        ActionProgress()
                            .accessibilityHidden(true)
                    }
                }
                .frame(maxWidth: size == .large ? .infinity : nil, minHeight: inlineLink ? nil : 24)
                .padding(.horizontal, inlineLink ? 0 : EnteSpacing.xl)
                .padding(.vertical, inlineLink ? 4 : density == .regular ? 14 : 12)
        }
        .buttonStyle(
            ActionStyle(
                variant: variant,
                enabled: enabled,
                cornerRadius: inlineLink ? 0 : EnteRadius.button
            )
        )
        .disabled(loading)
        .accessibilityLabel(title)
    }
}

private struct ActionStyle: ButtonStyle {
    @Environment(\.entePalette) private var palette
    let variant: ActionVariant
    let enabled: Bool
    let cornerRadius: CGFloat

    func makeBody(configuration: Configuration) -> some View {
        let colors = colors(pressed: configuration.isPressed)
        return configuration.label
            .foregroundStyle(colors.foreground)
            .background(colors.background)
            .clipShape(RoundedRectangle(cornerRadius: cornerRadius, style: .continuous))
            .scaleEffect(configuration.isPressed ? 0.98 : 1)
            .animation(.easeOut(duration: EnteMotion.quick), value: configuration.isPressed)
    }

    private func colors(pressed: Bool) -> (background: Color, foreground: Color) {
        guard enabled else {
            return (
                variant == .criticalText || variant == .link ? .clear : palette.fill,
                palette.disabledText
            )
        }
        return switch variant {
        case .primary: (pressed ? palette.primaryDarker : palette.primary, .white)
        case .secondary: (pressed ? palette.fillDarkest : palette.fill, palette.text)
        case .neutral: (palette.text, palette.reverseText)
        case .critical: (pressed ? palette.dangerDarker : palette.danger, .white)
        case .criticalText: (.clear, pressed ? palette.dangerDarker : palette.danger)
        case .link: (.clear, pressed ? palette.primaryDarker : palette.primary)
        }
    }
}

private struct ActionProgress: View {
    @State private var rotating = false

    var body: some View {
        Circle()
            .trim(from: 0.15, to: 0.85)
            .stroke(style: StrokeStyle(lineWidth: 2, lineCap: .round))
            .frame(width: 20, height: 20)
            .rotationEffect(.degrees(rotating ? 360 : 0))
            .animation(.linear(duration: 0.8).repeatForever(autoreverses: false), value: rotating)
            .onAppear { rotating = true }
    }
}

File: android/packages/components/src/main/kotlin/io/ente/components/ActionButton.kt
package io.ente.components

import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.tween
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.interaction.collectIsPressedAsState
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.style.TextDecoration
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp

public enum class ActionVariant {
    Primary,
    Secondary,
    Neutral,
    Critical,
    CriticalText,
    Link,
}

public enum class ActionDensity {
    Regular,
    Compact,
}

public enum class ActionSize {
    Small,
    Large,
}

@Composable
public fun ActionButton(
    label: String,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    variant: ActionVariant = ActionVariant.Primary,
    size: ActionSize = ActionSize.Large,
    density: ActionDensity = ActionDensity.Regular,
    enabled: Boolean = true,
    loading: Boolean = false,
) {
    val palette = LocalEntePalette.current
    val interactionSource = remember { MutableInteractionSource() }
    val pressed by interactionSource.collectIsPressedAsState()
    val colors = actionColors(variant, palette, enabled, pressed && !loading)
    val scale by
        animateFloatAsState(
            targetValue = if (pressed && enabled && !loading) 0.98f else 1f,
            animationSpec = tween(EnteMotion.quick),
            label = "actionScale",
        )
    val inlineLink = variant == ActionVariant.Link && size == ActionSize.Small
    val actionModifier = if (size == ActionSize.Large) modifier.fillMaxWidth() else modifier

    Surface(
        color = colors.background,
        contentColor = colors.foreground,
        shape = RoundedCornerShape(if (inlineLink) 0.dp else EnteRadius.button),
        modifier =
            actionModifier
                .semantics { contentDescription = label }
                .graphicsLayer {
                    scaleX = scale
                    scaleY = scale
                }
                .clickable(
                    enabled = enabled && !loading,
                    role = Role.Button,
                    interactionSource = interactionSource,
                    indication = null,
                    onClick = onClick,
                ),
    ) {
        Box(
            modifier =
                Modifier.padding(
                        horizontal = if (inlineLink) 0.dp else EnteSpacing.xl,
                        vertical =
                            if (inlineLink) {
                                4.dp
                            } else if (density == ActionDensity.Regular) {
                                14.dp
                            } else {
                                12.dp
                            },
                    )
                    .heightIn(min = if (inlineLink) 0.dp else 24.dp),
            contentAlignment = Alignment.Center,
        ) {
            Text(
                text = label,
                modifier = Modifier.alpha(if (loading) 0f else 1f).clearAndSetSemantics {},
                maxLines = 2,
                overflow = TextOverflow.Ellipsis,
                style =
                    if (density == ActionDensity.Regular) {
                        EnteTypography.bodyBold
                    } else {
                        EnteTypography.body
                    },
                textDecoration =
                    if (variant == ActionVariant.CriticalText || variant == ActionVariant.Link) {
                        TextDecoration.Underline
                    } else {
                        null
                    },
            )
            if (loading) {
                CircularProgressIndicator(
                    modifier = Modifier.size(20.dp).clearAndSetSemantics {},
                    color = colors.foreground,
                    strokeWidth = 2.dp,
                )
            }
        }
    }
}

private data class ActionColors(val background: Color, val foreground: Color)

private fun actionColors(
    variant: ActionVariant,
    palette: Palette,
    enabled: Boolean,
    pressed: Boolean,
): ActionColors {
    if (!enabled) {
        val background =
            if (variant == ActionVariant.CriticalText || variant == ActionVariant.Link) {
                Color.Transparent
            } else {
                palette.fill
            }
        return ActionColors(background, palette.disabledText)
    }
    return when (variant) {
        ActionVariant.Primary ->
            ActionColors(if (pressed) palette.primaryDarker else palette.primary, Color.White)
        ActionVariant.Secondary ->
            ActionColors(if (pressed) palette.fillDarkest else palette.fill, palette.text)
        ActionVariant.Neutral -> ActionColors(palette.text, palette.reverseText)
        ActionVariant.Critical ->
            ActionColors(if (pressed) palette.dangerDarker else palette.danger, Color.White)
        ActionVariant.CriticalText ->
            ActionColors(Color.Transparent, if (pressed) palette.dangerDarker else palette.danger)
        ActionVariant.Link ->
            ActionColors(Color.Transparent, if (pressed) palette.primaryDarker else palette.primary)
    }
}
